### PR TITLE
Feature/931 detect transaction

### PR DESCRIPTION
**Pull request by sadiabbasi**

### CODE REVIEW

File: src/containers/MainContainer.tsx [hunk 1]
```diff
@@ -103,6 +115,74 @@ export default function MainContainer({ did }: { did?: string }) {
...
+
+                try {
+                    const approvedNamespaces = buildApprovedNamespaces({
+                        proposal: proposal.params,
+                        supportedNamespaces: {
+                            eip155: {
+                                chains: ['eip155:11155111'], //11155111
+                                methods: ['eth_sendTransaction', 'personal_sign'],
+                                events: ['accountsChanged', 'chainChanged'],
+                                accounts: ['eip155:11155111:0x253c8d99c27d47A4DcdB04B40115AB1dAc466280'],
```
Comment (sadiabbasi) [inline src/containers/MainContainer.tsx:148]: @theblockstalk how we can get this address dynamically? as if we are getting using web3 it also require private key and return address

File: src/config/config.production.json [hunk 2]
```diff
@@ -39,5 +39,7 @@
     "loggerLevel": "error",
     "captchaSiteKey": "0caf1023-edb9-466c-b848-6f2e0fe975eb",
     "blockExplorerUrl": "https://explorer.pangea.web4.world",
-    "tonomyIdSlug": "united-wallet://"
+    "tonomyIdSlug": "united-wallet://",
+    "walletProjectId": ""
```
Comment (theblockstalk) [inline src/config/config.production.json:43]: call this walletConnectProjectId for clarity. Tonomy is a wallet so might be Confusing

File: src/containers/MainContainer.tsx [hunk 3]
```diff
@@ -103,6 +115,74 @@ export default function MainContainer({ did }: { did?: string }) {
...
+        try {
+            console.log('walletconnect');
+            const core = new Core({
+                projectId: settings.config.walletProjectId,
+            });
+
+            const web3wallet = await Web3Wallet.init({
+                core,
+                metadata: {
+                    name: 'Demo app',
```
Comment (theblockstalk) [inline src/containers/MainContainer.tsx:128]: What is this data about? does this need too come from the app or is this about the Tonomy app, in which case should use information from the settings probably...

File: src/containers/MainContainer.tsx [hunk 4]
```diff
@@ -103,6 +115,74 @@ export default function MainContainer({ did }: { did?: string }) {
...
+            web3wallet.on('session_proposal', async (proposal) => {
+                console.log('proposal', proposal.params);
+                console.log(proposal.params.proposer.metadata);
+
+                try {
+                    const approvedNamespaces = buildApprovedNamespaces({
+                        proposal: proposal.params,
+                        supportedNamespaces: {
+                            eip155: {
+                                chains: ['eip155:11155111'], //11155111
```
Comment (theblockstalk) [inline src/containers/MainContainer.tsx:145]: is this sepolia? sepolia should be allowed for development/testnet, and production should use Ethreum mainnet

File: src/containers/MainContainer.tsx (hunk 1 above)
Comment (theblockstalk) [reply src/containers/MainContainer.tsx:148]: Is this a required field? It is not part of the information sent in the login request?

IMO You are correct - this should be set dynamically based on the users account. I'm not sure how

File: src/containers/MainContainer.tsx [hunk 5]
```diff
@@ -103,6 +115,74 @@ export default function MainContainer({ did }: { did?: string }) {
...
+                        id: proposal.id,
+                        reason: getSdkError('USER_REJECTED'),
+                    });
+                }
+            });
+
+            try {
+                web3wallet.on('session_request', async (event) => {
+                    console.log('event', event?.params?.request?.method);
+                    if (event?.params?.request?.method === 'eth_sendTransaction')
```
Comment (theblockstalk) [inline src/containers/MainContainer.tsx:170]: you support `personal_sign` method as well on L146. How is this handled?

File: src/containers/MainContainer.tsx (hunk 4 above)
Comment (sadiabbasi) [reply src/containers/MainContainer.tsx:145]: yess i will make this dynamic..

File: src/utils/Web3WalletClient.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    core = new Core({
+        // @notice: If you want the debugger / logs
+        // logger: 'debug',
+        projectId: '2850896ad9cf6c1d958203b00b199c2d',
+        relayUrl: 'wss://relay.walletconnect.com',
+    });
+
+    const { eip155Addresses } = await createOrRestoreEIP155Wallet();
+
+    currentETHAddress = '0x253c8d99c27d47A4DcdB04B40115AB1dAc466280'; //eip155Addresses[0];
```
Comment (sadiabbasi) [inline src/utils/Web3WalletClient.ts:24]: currently i added address for test

File: src/containers/MainContainer.tsx [hunk 7]
```diff
@@ -103,10 +132,79 @@ export default function MainContainer({ did }: { did?: string }) {
...
+
+    const onSessionRequest = useCallback(async (requestEvent: SignClientTypes.EventArguments['session_request']) => {
+        const { topic, params } = requestEvent;
+        const { request } = params;
+        const requestSessionData = web3wallet.engine.signClient.session.get(topic);
+
+        console.log('request', request.method);
+
+        switch (request.method) {
+            case EIP155_SIGNING_METHODS.PERSONAL_SIGN:
```
Comment (theblockstalk) [inline src/containers/MainContainer.tsx:147]: what is personal sign?

File: src/containers/MainContainer.tsx [hunk 8]
```diff
@@ -103,10 +132,79 @@ export default function MainContainer({ did }: { did?: string }) {
...
+                return;
+            case EIP155_SIGNING_METHODS.ETH_SEND_TRANSACTION:
+                setRequestSession(requestSessionData);
+                setRequestEventData(requestEvent);
+                setTransactionModal(true);
```
Comment (theblockstalk) [inline src/containers/MainContainer.tsx:152]: this is a modal or a new screen?

File: src/containers/MainContainer.tsx [hunk 9]
```diff
@@ -103,10 +132,79 @@ export default function MainContainer({ did }: { did?: string }) {
         }
     }
 
+    const onSessionProposal = useCallback((proposal: SignClientTypes.EventArguments['session_proposal']) => {
+        setPairedProposal(proposal);
+    }, []);
+
+    const onSessionRequest = useCallback(async (requestEvent: SignClientTypes.EventArguments['session_request']) => {
```
Comment (theblockstalk) [inline src/containers/MainContainer.tsx:139]: can we set this up in the Communication container? This is where we wait for the other requests. Keeps all logic in the same place...

File: src/containers/MainContainer.tsx (hunk 8 above)
Comment (sadiabbasi) [reply src/containers/MainContainer.tsx:152]: for testing i added modal to check

File: src/config/config.json [hunk 10]
```diff
@@ -29,7 +29,7 @@
         "transparencyLearnMore": "#",
         "securityLearnMore": "#"
     },
-    "expoProjectId": "6c32ba89-729d-4286-b909-e1f48918839e",
+    "expoProjectId": "3a65b31c-572d-4537-9ead-f7c4bddf7d29",
```
[the remark this answers is not in the document]
Comment (sadiabbasi) [reply src/config/config.json:32]: before it was not correct

File: src/utils/EIP155Request.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import { providers } from 'ethers';
+import { currentETHAddress } from '../utils/Web3WalletClient';
+
+export async function approveEIP155Request(requestEvent: SignClientTypes.EventArguments['session_request']) {
+    const { params, id } = requestEvent;
+    const { chainId, request } = params;
+    const wallet = eip155Wallets[getWalletAddressFromParams([currentETHAddress], params)];
+
+    switch (request.method) {
+        case EIP155_SIGNING_METHODS.PERSONAL_SIGN: {
```
[the remark this answers is not in the document]
Comment (sadiabbasi) [reply src/utils/EIP155Request.ts]: https://docs.walletconnect.com/advanced/multichain/rpc-reference/ethereum-rpc#personal_sign 
https://docs.walletconnect.com/advanced/multichain/rpc-reference/ethereum-rpc#eth_sendtransaction

File: src/containers/LoginPassphraseContainer.tsx [hunk 12]
```diff
@@ -35,10 +39,32 @@ export default function LoginPassphraseContainer({
         await user.updateKeys(passphrase.join(' '));
     }
 
+    async function createEthereumAccount() {
```
Comment (theblockstalk) [inline src/containers/LoginPassphraseContainer.tsx:42]: remove the implementation from `src/containers/HcaptchaContainer.tsx` lets make a common function that does this for both create account and login.

1. take passphrase
2. call `generateSeedFromPassword()` to get seed
3. then call `generatePrivateKeyFromSeed()` to create the key for Ethereum (and other chains)

this should all be done without blocking user from progress (aka happens in background without `await`)

File: src/hooks/useWalletConnect.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    const user = userStore.user;
+
+    const onInitialize = useCallback(async () => {
+        try {
+            const u = await user.getUsername();
+
+            const username = u.getBaseUsername();
+
+            if (username) {
+                const key = await generatePrivateKeyForEthereum(getPassphrase(), username); //TODO update this
```
Comment (theblockstalk) [inline src/hooks/useWalletConnect.tsx:20]: this needs to be done at create account or login, not also here.

File: src/services/WalletConnect/EIP155MethodsAndChain.tsx [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+ */
+export type TEIP155Chain = keyof typeof EIP155_CHAINS;
+
+/**
+ * Chains
+ */
+export const EIP155_MAINNET_CHAINS = {
+    'eip155:1': {
+        chainId: 1,
+        name: 'Ethereum',
```
Comment (theblockstalk) [inline src/services/WalletConnect/EIP155MethodsAndChain.tsx:17]: use our new EthereumChain class to fill this

File: src/services/WalletConnect/EIP155Request.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+import { currentETHAddress } from './Web3WalletClient';
+import { ethers } from 'ethers';
+
+export async function approveEIP155Request(requestEvent: SignClientTypes.EventArguments['session_request']) {
+    const { params, id } = requestEvent;
+    const { chainId, request } = params;
+
+    switch (request.method) {
+        case EIP155_SIGNING_METHODS.ETH_SEND_TRANSACTION: {
+            // const provider = new providers.JsonRpcProvider(EIP155_CHAINS[chainId as TEIP155Chain].rpc);
```
Comment (theblockstalk) [inline src/services/WalletConnect/EIP155Request.ts:15]: comments?

File: src/veramo/setup.ts [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+const dbConnection = new DataSource({
+    type: 'expo',
+    driver: require('expo-sqlite'),
+    database: 'veramo.sqlite',
+    logging: ['error', 'info', 'warn'],
+}).initialize();
+
+export const agent = createAgent<IDIDManager & IKeyManager & IDataStore & IDataStoreORM>({
+    plugins: [
```
Comment (theblockstalk) [inline src/veramo/setup.ts:17]: can we just use the KeyManager object directly, without creating an agent? Might be cleaner... But maybe not.

File: src/veramo/setup.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+import { KeyManager } from '@veramo/key-manager';
+import { KeyManagementSystem, SecretBox } from '@veramo/kms-local';
+import { KeyStore, PrivateKeyStore } from '@veramo/data-store';
+import { DataSource } from 'typeorm';
+
+const DB_ENCRYPTION_KEY = '29739248cad1bd1a0fc4d9b75cd4d2990de535baf5caadfdf8d8f86664aa830c';
+
+const dbConnection = new DataSource({
+    type: 'expo',
+    driver: require('expo-sqlite'),
```
Comment (theblockstalk) [inline src/veramo/setup.ts:11]: better to use `import` at start of file

File: src/veramo/setup.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
+import { createAgent, IDataStore, IDataStoreORM, IKeyManager, IDIDManager } from '@veramo/core';
+import { KeyManager } from '@veramo/key-manager';
+import { KeyManagementSystem, SecretBox } from '@veramo/kms-local';
+import { KeyStore, PrivateKeyStore } from '@veramo/data-store';
+import { DataSource } from 'typeorm';
+
+const DB_ENCRYPTION_KEY = '29739248cad1bd1a0fc4d9b75cd4d2990de535baf5caadfdf8d8f86664aa830c';
```
Comment (theblockstalk) [inline src/veramo/setup.ts:7]: should be environment variable.

later on, this should be the user fingerprint or PIN unlocked secret

File: src/containers/HcaptchaContainer.tsx [hunk 19]
```diff
@@ -75,13 +78,31 @@ export default function HcaptchaContainer({ navigation }: { navigation: Props['n
...
         } catch (e) {
             errorStore.setError({ error: e, expected: false });
         }
     }
 
+    async function createEthereumAccount() {
+        const key = await generatePrivateKeyFromPassword(getPassphrase());
+        const accountName = (await user.getAccountName()).toString();
+
+        await agent.keyManagerImportKey({
```
[the remark this answers is not in the document]
Comment (sadiabbasi) [reply src/containers/HcaptchaContainer.tsx:94]: i just tested now and getting the same error..

**Comment by sadiabbasi** [comment]
split this in 2 separate tasks

**Pull request closed without merging**